File: prettier-plugin-csharpier/src/Printer/Types/DoStatement.ts
import { PrintMethod } from "../PrintMethod";
import { SyntaxToken, printPathSyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";

export interface DoStatementNode extends SyntaxTreeNode<"DoStatement"> {
    doKeyword: SyntaxToken;
    statement: SyntaxTreeNode;
    whileKeyword: SyntaxToken;
    condition: SyntaxTreeNode;
}

export const printDoStatement: PrintMethod<DoStatementNode> = (path, options, print) => {
    return concat([
        printPathSyntaxToken(path, "doKeyword"),
        path.call(print, "statement"),
        hardline,
        printPathSyntaxToken(path, "whileKeyword"),
        " (",
        path.call(print, "condition"),
        ");",
    ]);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/DoStatement.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class DoStatement
{
    public static Doc Print(DoStatementSyntax node, FormattingContext context)
    {
        return Doc.Concat(
            ExtraNewLines.Print(node),
            Token.PrintWithSuffix(
                node.DoKeyword,
                node.Statement is not BlockSyntax ? " " : Doc.Null,
                context
            ),
            Node.Print(node.Statement, context),
            node.Statement is BlockSyntax ? " " : Doc.HardLine,
            Token.PrintWithSuffix(node.WhileKeyword, " ", context),
            Token.Print(node.OpenParenToken, context),
            Doc.Group(Doc.Indent(Doc.SoftLine, Node.Print(node.Condition, context)), Doc.SoftLine),
            Token.Print(node.CloseParenToken, context),
            Token.Print(node.SemicolonToken, context)
        );
    }
}
